rrmngmnt/nmcli.py

```diff
@@ -0,0 +1,644 @@
+                    "dev show {con}"
+                ).format(con=name)
+            )
+            properties = out.splitlines()
+            devices.append(
+                {
+                    "name": properties[0],
+                    "type": properties[1],
+                    "mac": properties[2],
+                    "mtu": properties[3],
+                }
+            )
+
+        return devices
+
+    def get_device_type(self, device):
+        """
+        Gets the device type.
+
+        Args:
+            device (str): device name.
+
+        Returns:
+            str: the device type.
+
+        Raises:
+            CommandExecutionFailure: if the remote host returned a code
+            indicating a failure in execution.
+        """
+        return self._exec_command(
+            command="nmcli -g GENERAL.TYPE device show {dev}".format(dev=device)  # noqa: E501
+        ).strip()
+
+    def set_connection_state(self, connection, state):
+        """
+        Sets a connection's state.
+
+        Args:
+            connection (str): name, UUID or path.
+            state (str): the desired state.
+                available states are: ["up", "down"].
+
+        Raises:
+            CommandExecutionFailure: if the remote host returned a code
+            indicating a failure in execution.
+        """
+        self._exec_command(
+            command="nmcli connection {state} {con}".format(state=state, con=connection)  # noqa: E501
+        )
+
+    def add_ethernet_connection(
+        self,
+        con_name,
+        ifname,
+        auto_connect=None,
+        save=None,
+        mac=None,
+        mtu=None,
+        ipv4_method=None,
+        ipv4_addr=None,
+        ipv4_gw=None,
+        ipv6_method=None,
+        ipv6_addr=None,
+        ipv6_gw=None,
+    ):
+        """
+        Creates an ETHERNET connection.
+
+        Args:
+            con_name (str): the created connection's name.
+            ifname (str): the interface name to use.
+            auto_connect (bool): True to connect automatically, or False for
+                manual.
+            save (bool): True to persist the connection, or False.
+            mac (str): MAC address to set for the connection.
+            mtu (int): MTU to set for the connection.
+            ipv4_method (str): setting method.
+                Available methods: auto, disabled, link-local, manual, shared.
+            ipv4_addr (str): a static address.
+            ipv4_gw (str): a gateway address.
+            ipv6_method (str): setting method.
+                Available methods: auto, dhcp, disabled, ignore, link-local,
+                manual, shared.
+            ipv6_addr (str): a static address.
+            ipv6_gw (str): a gateway address.
+
+        Raises:
+            CommandExecutionFailure: if the remote host returned a code
+                indicating a failure in execution.
+        """
+        type_options = {}
```

Comment:
what is type_option?
I guess this is connection_params or similar.

Reply:
Each connection type has its own specific options.
A bond, for example, has a mode option. a VLAN has a VLAN ID etc..
I'm building these and passing them to _nmcli_con_cmd_builder so he will assemble them into the command where they belong.